packages/connect-node/src/connect-http-transport.ts

```diff
@@ -0,0 +1,353 @@
+
+const messageFlag = 0b00000000;
+
+export interface ConnectHttpTransportOptions {
+  /**
+   * Base URI for all HTTP requests.
+   *
+   * Requests will be made to <baseUrl>/<package>.<service>/method
+   *
+   * Example: `baseUrl: "http://example.com/my-api"`
+   *
+   * This will make a `POST /my-api/my_package.MyService/Foo` to
+   * `example.com` via HTTPS.
+   */
+  baseUrl: string;
+
+  /**
+   * By default, clients use the binary format for gRPC-web, because
+   * not all gRPC-web implementations support JSON.
+   */
+  useBinaryFormat?: boolean;
+
+  /**
+   * Interceptors that should be applied to all calls running through
+   * this transport. See the Interceptor type for details.
+   */
+  interceptors?: Interceptor[];
+
+  /**
+   * Options for the JSON format.
+   */
+  jsonOptions?: Partial<JsonReadOptions & JsonWriteOptions>;
+
+  /**
+   * Options for the binary wire format.
+   */
+  binaryOptions?: Partial<BinaryReadOptions & BinaryWriteOptions>;
+}
+
+/**
+ * Create a Transport for the connect protocol on Http.
+ *
+ */
+export function createConnectHttpTransport(
+  options: ConnectHttpTransportOptions
+): Transport {
+  const useBinaryFormat = options.useBinaryFormat ?? false;
+  return {
+    async unary<
+      I extends Message<I> = AnyMessage,
+      O extends Message<O> = AnyMessage
+    >(
+      service: ServiceType,
+      method: MethodInfo<I, O>,
+      signal: AbortSignal | undefined,
+      timeoutMs: number | undefined,
+      header: HeadersInit | undefined,
+      message: PartialMessage<I>
+    ): Promise<UnaryResponse<O>> {
+      try {
+        const { normalize, parse } = createClientMethodSerializers(
+          method,
+          useBinaryFormat,
+          options.jsonOptions,
+          options.binaryOptions
+        );
+        const createRequestHeader = connectCreateRequestHeader.bind(
+          null,
+          method.kind,
+          useBinaryFormat
+        );
+
+        const validateContentType = connectExpectContentType.bind(
+          null,
+          method.kind,
+          useBinaryFormat
+        );
+
+        return await runUnary<I, O>(
+          {
+            stream: false,
+            service,
+            method,
+            url: createMethodUrl(options.baseUrl, service, method).toString(),
+            init: {},
+            header: createRequestHeader(timeoutMs, header),
+            message: normalize(message),
+            signal: signal ?? new AbortController().signal,
+          },
+          async (req: UnaryRequest<I>): Promise<UnaryResponse<O>> => {
+            const response = await nodeRequest({
+              req,
+              useBinaryFormat,
+              jsonOptions: options.jsonOptions,
+              binaryOptions: options.binaryOptions,
+            });
+
+            const responseHeaders = nodeHeaderToWebHeader(response.headers);
+            await validateResponseHeader(
+              response.statusCode as number,
```

Comment:
unfortunately the type for `statusCode` is `IncomingMessage.statusCode?: number | undefined` which is why i casted. I could change it up to check if we have a status code or default to 500?

Reply:
I think that is because `http.IncomingMessage` is used both for client responses (where it should always be present, because a HTTP/1 response always has a status), and for incoming server requests (where it cannot be present by definition).

Since we are on the client side here, I think it is fair that we error on a response that cannot be valid. You can use private/assert.ts for this:

```ts
assert(typeof response.statusCode == "number", "http1 client response is missing status code");
```